contracts/Gap.sol

```diff
@@ -4,34 +4,117 @@ pragma solidity 0.8.19;
 
 import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
 import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
-import {IEAS, AttestationRequest, AttestationRequestData, MultiAttestationRequest} from "@ethereum-attestation-service/eas-contracts/contracts/IEAS.sol";
+import {IEAS, Attestation, AttestationRequest, AttestationRequestData, MultiAttestationRequest, MultiRevocationRequest} from "@ethereum-attestation-service/eas-contracts/contracts/IEAS.sol";
 
 contract Gap is Initializable, OwnableUpgradeable {
     IEAS public eas;
 
     struct AttestationRequestNode {
         bytes32 uid;
         MultiAttestationRequest multiRequest;
-        uint refIdx;
+        uint256 refIdx;
     }
 
     function initialize(address easAddr) public initializer {
         eas = IEAS(easAddr);
         __Ownable_init();
     }
 
-    function multiSequentialAttest(AttestationRequestNode[] calldata requestNodes) public {
+    /// 
+    /// Verify if msg.sender owns the referenced attestation
+    /// 
+    function canAttestToRef(bytes32 uid) public view {
+        Attestation memory ref = eas.getAttestation(uid);
+        require(
+            ref.attester == msg.sender || ref.recipient == msg.sender,
+            /* || msg.sender == owner() */
+            // ↑ Should owner be able to revoke any attestation?
+            "Not owner."
+        );
+    }
+
+    ///
+    /// Verify if msg.sender owns the set of attestations
+    ///
+    function canAttestToRefs(AttestationRequestData[] memory datas)
+        private
+        view
+    {
+        for (uint256 j = 0; j < datas.length; j++) {
+            if (datas[j].refUID != bytes32(0)) {
+                canAttestToRef(datas[j].refUID);
+            }
+        }
+    }
+
+    /// 
+    /// Revokes multiple attestations
+    /// 
+    function multiRevoke(MultiRevocationRequest[] calldata multiRequests)
+        external
+        payable
+    {
+        // Checks if every revoke request belongs to the sender
+        // The sender can be either the attester or the recipient.
+        for (uint256 i = 0; i < multiRequests.length; i++) {
+            MultiRevocationRequest memory chunk = multiRequests[i];
+            for (uint256 j = 0; j < chunk.data.length; j++) {
+                Attestation memory target = eas.getAttestation(
+                    chunk.data[j].uid
+                );
+
+                require(
+                    target.attester == msg.sender ||
+                        target.recipient == msg.sender,
+                    /* || msg.sender == owner()*/
+                    // ↑ Should owner be able to revoke any attestation?
+                    "Not owner."
+                );
+            }
+        }
+        eas.multiRevoke(multiRequests);
+    }
+
+    function attest(AttestationRequest calldata request)
```

Comment:
Single attestation is needed because the multi attest function args are very different from the original attestation, then we'll need to do several changes in the sdk, in this case we can use the same argument of the EAS contract and still use our verification. 
Single revocation is not needed because the arguments are the same so we don't need to change anything in the sdk